

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Taxonomic reference database preprocessing &mdash; RSP4ABM v.0.9.15 documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Pipeline execution" href="execution.html" />
    <link rel="prev" title="Setup and system requirements" href="setup.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home" alt="Documentation Home"> RSP4ABM
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="under_the_hood.html">Under the hood</a></li>
<li class="toctree-l1"><a class="reference internal" href="setup.html">Setup and system requirements</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Taxonomic reference database preprocessing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#rational">Rational:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#working-principle">Working principle:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#execution">Execution:</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#config-file">Config file</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#references">References:</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="execution.html">Pipeline execution</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">Frequently asked questions (FAQ)</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">RSP4ABM</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Taxonomic reference database preprocessing</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/pages/ref_DB_preprocessing.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="taxonomic-reference-database-preprocessing">
<h1>Taxonomic reference database preprocessing<a class="headerlink" href="#taxonomic-reference-database-preprocessing" title="Permalink to this headline">¶</a></h1>
<div class="section" id="rational">
<h2>Rational:<a class="headerlink" href="#rational" title="Permalink to this headline">¶</a></h2>
<p>After processing of the reads by a metagenomic pipeline, we expect amplicon sequences (OTUs or ASVs) to be assigned to the lowest possible taxonomic level (Species). However, all species cannot be differentiated without ambiguities based on short amplicon sequences. Thus, it is expected for some species to have more or less exactly the same sequence on the studied gene, and even more on the short fragment amplified and sequenced in amplicon-based metagenomics. Furthermore, the classifiers included in RSP4ABM (original RDP <a class="footnote-reference brackets" href="#id6" id="id1">1</a>, RDP integrated in Qiime <a class="footnote-reference brackets" href="#id7" id="id2">2</a> and Decipher IDTAXA <a class="footnote-reference brackets" href="#id8" id="id3">3</a>) all have specific formatting requirements and the two last require an initial training. Thus, to improve the taxonomic classification and to adapt the format of the provided reference database to these multiple classifiers, RSP4ABM includes a dedicated reference database preprocessing workflow.</p>
</div>
<div class="section" id="working-principle">
<h2>Working principle:<a class="headerlink" href="#working-principle" title="Permalink to this headline">¶</a></h2>
<p>The user indicates to the pipeline the sequences of the used PCR primers and the path to the reference database fasta file and taxonomy annotation file to be formatted.</p>
<p>With that and using tools from Qiime2 <a class="footnote-reference brackets" href="#id9" id="id4">4</a> and VSEARCH <a class="footnote-reference brackets" href="#id10" id="id5">5</a>, as well as home-made scripts, the pipeline will first extract the amplicon matching the used primes. Then, it unifies the taxonomy: in cases where the exact same amplicon is predicted for multiple taxa, it collapses together their identifiers at the genus/species (up to a user-defined number of occurrences). An error is raised in cases where the same sequence is observed across different families.</p>
<p>The original RDP classier as well as Decipher IDTAXA require specific formatting and pre-training of the classifiers. These steps are automatically conducted by RSP4ABM when running this preprocessing step.</p>
<p>Finally, the  pipeline will make a copy of the original database as well as computes hashes of all files for traceability purposes.</p>
</div>
<div class="section" id="execution">
<h2>Execution:<a class="headerlink" href="#execution" title="Permalink to this headline">¶</a></h2>
<p>A dedicated workflow is embedded in RSP4ABM for database preprocessing. This workflow is to be run only one time for each set of PCR primer and reference database.</p>
<div class="section" id="config-file">
<h3>Config file<a class="headerlink" href="#config-file" title="Permalink to this headline">¶</a></h3>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1">## Localisation of input database</span>
<span class="nt">DBpath_seq</span><span class="p">:</span> <span class="s">&quot;/home/master_students/bin/microbiome16S_pipeline/data/ezbiocloud201805/DB_amp.fasta&quot;</span> <span class="c1"># reference genomic sequence in .fasta file</span>
<span class="nt">DBpath_tax</span><span class="p">:</span> <span class="s">&quot;/home/master_students/bin/microbiome16S_pipeline/data/ezbiocloud201805/DB_amp_taxonomy.txt&quot;</span> <span class="c1"># reference taxonomic of these sequence</span>

<span class="c1">## Output</span>
<span class="nt">tax_DB_path</span><span class="p">:</span> <span class="s">&quot;/data/databases/amplicon_based_metagenomics/16S/&quot;</span> <span class="c1"># path where to write the output of the database processing pipeline. </span>
<span class="nt">tax_DB_name</span><span class="p">:</span> <span class="s">&quot;EzBioCloud_version&quot;</span> <span class="c1"># The desired name of processed pipeline. </span>

<span class="c1">#trimming</span>
<span class="nt">forward_primer</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">CCTACGGGNGGCWGCAG</span> <span class="c1"># the forward PCR primer used to extract the amplicon</span>
<span class="nt">reverse_primer</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">GACTACHVGGGTATCTAATCC</span> <span class="c1"># the reverse PCR primer used to extract the amplicon</span>
<span class="nt">length_max</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">500</span> <span class="c1"># the max length of the amplicon</span>

<span class="c1">#Filtering</span>
<span class="nt">numbers_species</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">4</span> <span class="c1"># the max number of species names to be pasted together </span>
<span class="nt">numbers_genus</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">2</span> <span class="c1"># the max number of genus names to be pasted together </span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="references">
<h2>References:<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<dl class="footnote brackets">
<dt class="label" id="id6"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Wang Q, Garrity GM, Tiedje JM, Cole JR. Naïve Bayesian classifier for rapid assignment of rRNA sequences into the new bacterial taxonomy. Appl Environ Microbiol. 2007.</p>
</dd>
<dt class="label" id="id7"><span class="brackets"><a class="fn-backref" href="#id2">2</a></span></dt>
<dd><p>Caporaso JG, Kuczynski J, Stombaugh J, Bittinger K, Bushman FD, Costello EK, et al. QIIME allows analysis of high-throughput community sequencing data. Nature Methods. 2010.</p>
</dd>
<dt class="label" id="id8"><span class="brackets"><a class="fn-backref" href="#id3">3</a></span></dt>
<dd><p>Murali A, Bhargava A, Wright ES. IDTAXA: A novel approach for accurate taxonomic classification of microbiome sequences. Microbiome. 2018.</p>
</dd>
<dt class="label" id="id9"><span class="brackets"><a class="fn-backref" href="#id4">4</a></span></dt>
<dd><p>Estaki M, Jiang L, Bokulich NA, McDonald D, González A, Kosciolek T, et al. QIIME 2 Enables Comprehensive End-to-End Analysis of Diverse Microbiome Data and Comparative Studies with Publicly Available Data. Curr Protoc Bioinforma. 2020.</p>
</dd>
<dt class="label" id="id10"><span class="brackets"><a class="fn-backref" href="#id5">5</a></span></dt>
<dd><p>Rognes T, Flouri T, Nichols B, Quince C, Mahé F. VSEARCH: A versatile open source tool for metagenomics. PeerJ. 2016</p>
</dd>
</dl>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="execution.html" class="btn btn-neutral float-right" title="Pipeline execution" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="setup.html" class="btn btn-neutral float-left" title="Setup and system requirements" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Valentin Scherz

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>